wakuv2/filter_manager.go

```diff
@@ -53,225 +28,95 @@ type FilterEvent struct {
-	switch ev.eventType {
+func (mgr *FilterManager) filterLifecycle(ctx context.Context, filterID string, contentFilter protocol.ContentFilter) {
+	logger := mgr.logger.With(zap.String("filterID", filterID), zap.Any("contentFilter", contentFilter))
 
-	case FilterEventAdded:
-		mgr.filterSubs[ev.filterID] = make(subscription.SubscriptionSet)
-		mgr.resubscribe(ev.filterID)
-
-	case FilterEventRemoved:
-		for _, sub := range mgr.filterSubs[ev.filterID] {
-			if sub == nil {
-				// Skip temp subs
-				continue
-			}
-			go mgr.unsubscribeFromFilter(ev.filterID, sub)
-		}
-		delete(mgr.filterSubs, ev.filterID)
-
-	case FilterEventPingResult:
-		if ev.success {
-			break
-		}
-		// filterID field is only set when there are no subs to check for this filter,
-		// therefore no particular peers that could be unreachable.
-		if ev.filterID != "" {
-			// Trigger full resubscribe, filter has too few peers
-			mgr.logger.Debug("filter has too few subs", zap.String("filterId", ev.filterID))
-			mgr.resubscribe(ev.filterID)
-			break
-		}
+	// Use it to ping filter peer(s) periodically
+	ticker := time.NewTicker(5 * time.Second)
+	defer ticker.Stop()
 
-		// Delete subs for removed peer
-		for filterID, subs := range mgr.filterSubs {
-			for _, sub := range subs {
-				if sub == nil {
-					// Skip temp subs
-					continue
-				}
-				if sub.PeerID == ev.peerID {
-					mgr.logger.Debug("filter sub is inactive", zap.String("filterId", filterID), zap.Stringer("peerId", sub.PeerID), zap.String("subID", sub.ID))
-					delete(subs, sub.ID)
-					go mgr.unsubscribeFromFilter(filterID, sub)
-				}
-			}
-			mgr.resubscribe(filterID)
-		}
+	// A single sub for this filter
+	var sub *subscription.SubscriptionDetails
 
-	case FilterEventSubscribeResult:
-		subs, found := mgr.filterSubs[ev.filterID]
-		if ev.success {
-			if found {
-				subs[ev.sub.ID] = ev.sub
-				go mgr.runFilterSubscriptionLoop(ev.sub)
+	for {
+		logger.Debug("filter lifecycle started")
+		if sub != nil {
+			logger.Debug("filter health check")
+			err := mgr.isFilterSubAlive(ctx, sub)
+			if err != nil {
+				logger.Debug("filter ping error", zap.Error(err))
+				mgr.unsubscribe(ctx, logger, sub)
+				sub = nil
 			} else {
-				// We subscribed to a filter that is already uninstalled; invoke unsubscribe
-				go mgr.unsubscribeFromFilter(ev.filterID, ev.sub)
+				logger.Debug("filter health ok")
 			}
-		}
-		if found {
-			// Delete temp subscription record
-			delete(subs, ev.tempID)
+		} else {
+			logger.Debug("filter has no sub")
 		}
 
-	case FilterEventUnsubscribeResult:
-		mgr.logger.Debug("filter event unsubscribe result", zap.String("filterId", ev.filterID), zap.Stringer("peerID", ev.sub.PeerID))
-
-	case FilterEventGetStats:
-		stats := make(FilterSubs)
-		for id, subs := range mgr.filterSubs {
-			stats[id] = make(subscription.SubscriptionSet)
-			for subID, sub := range subs {
-				if sub == nil {
-					// Skip temp subs
-					continue
-				}
+		if sub == nil {
```

Comment:
In that case we are not going to attempt re-subscribing after failed ping. Suppose `sub != nil` when we enter the loop. We invoke `isFilterSubAlive` and it fails. In case when we merge the condition with `else`, we'll skip resubscribing until the next `ticker` event. The code will be correct, but there will be a delay.